applications/GeoMechanicsApplication/custom_processes/apply_c_phi_reduction_process.hpp

```diff
@@ -0,0 +1,204 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors: Marjan Fathian
+//                Wijtze Pieter Kikstra
+//                Anne van de Graaf
+//
+
+#pragma once
+
+// System includes
+#include <cmath>
+#include <iostream>
+
+// Project includes
+#include "custom_elements/small_strain_U_Pw_diff_order_element.hpp"
+#include "custom_utilities/stress_strain_utilities.h"
+#include "includes/element.h"
+#include "includes/model_part.h"
+#include "utilities/math_utils.h"
+
+// Application includes
+#include "geo_mechanics_application_variables.h"
+#include "includes/kratos_flags.h"
+#include "includes/kratos_parameters.h"
+#include "processes/process.h"
+
+namespace Kratos
+{
+
+class ApplyCPhiReductionProcess : public Process
+{
+public:
+    KRATOS_CLASS_POINTER_DEFINITION(ApplyCPhiReductionProcess);
+
+    ApplyCPhiReductionProcess(ModelPart& model_part, const Parameters&)
+        : Process(Flags()), mrModelPart(model_part)
+    {
+    }
+
+    ~ApplyCPhiReductionProcess() override = default;
+
+    void ExecuteInitializeSolutionStep() override
+    {
+        KRATOS_TRY
+
+        // Recognize whether a step is restarted. If so, try again with a smaller reduction increment.
+        if (mrModelPart.GetProcessInfo().GetValue(NUMBER_OF_CYCLES) > 1) mReductionIncrement *= 0.5;
+        mReductionFactor = mPreviousReductionFactor - mReductionIncrement;
+        KRATOS_INFO("ApplyCPhiReductionProces::ExecuteInitializeSolutionStep")
+            << "Try a c-phi reduction factor " << mReductionFactor << " (safety factor "
+            << 1. / mReductionFactor << ") Previous reduction = " << mPreviousReductionFactor
+            << " Reduction increment = " << mReductionIncrement << std::endl;
+
+        double    phi                = 0.;
+        double    reduced_phi        = 0.;
+        double    c                  = 0.;
+        double    reduced_c          = 0.;
+        IndexType previousPropertyId = -1;
+        // Apply C/Phi Reduction procedure for the model part:
+        block_for_each(mrModelPart.Elements(),
+                       [this, &phi, &reduced_phi, &c, &reduced_c, &previousPropertyId](Element& rElement) {
+            // Only compute new c and phi if the Id changes
+            if (mrModelPart.GetProperties(rElement.GetProperties().Id()).Id() != previousPropertyId) {
```

Comment:
This is a bit of a confusing line to me: is the Id() from the element different than the Id() of the properties in the model part? 

Here we now get the id of the properties of the element, then ask the model part which properties belong to it and then ask the id of these properties. Wouldn't this do the same?
```suggestion
            if (rElement.GetProperties().Id() != previousPropertyId) {
```

Reply:
It works with this simplification!